### PR TITLE
Arc - implement support for transitive interceptor bindings.

### LINKED ISSUE #2560: MP rest client interfaces do not support MP FT annotations
SmallRye FT implementation is using an additional interceptor binding `@HystrixCommandBinding` to bind `HystrixCommandInterceptor` to all bean clasess that declare an MP FT annotation (such as `@Fallback`). In order to make it work SM FT portable extension adds `@HystrixCommandBinding` to [all FT annotations](https://github.com/smallrye/smallrye-fault-tolerance/blob/master/implementation/src/main/java/io/smallrye/faulttolerance/HystrixExtension.java#L66-L71) and since interceptor bindings are transitive it works as expected.

However, quarkus does not support transitive interceptor bindings yet and so the `HystrixCommandInterceptor` is not bound to the following rest client interface:
```java
@RegisterRestClient
interface MyClient {

        @Fallback(MyFallback.class)
        @GET
        @Path("/data")
        JsonArray get();
    }
```
WORKAROUND: annotate the rest client with `@HystrixCommandBinding` (this annotation comes from SM FT).

I think we should support transitive interceptor bindings.

### PR DESCRIPTION
**Pull request by manovotn**
Provide tests for one/two levels deep bindings on both, methods and classes.

Resolves #2560

### CODE REVIEW

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 1]
```diff
@@ -361,6 +373,27 @@ void init() {
...
 
+    static Map<DotName, ClassInfo> findTransitiveInterceptorBindings(Collection<DotName> initialBindings,
+            IndexView index, Map<DotName, ClassInfo> result) {
+        // any annotation that contains interceptor binding annotation is effectively also an interceptor binding
+        for (DotName annotationName : initialBindings) {
+            Set<DotName> collectedAnnotations = new HashSet<>();
+            for (AnnotationInstance annotation : index.getAnnotations(annotationName)) {
+                if (Kind.CLASS.equals(annotation.target().kind())
+                        && Modifier.isInterface(annotation.target().asClass().flags())) {
+                    // TODO is this good way to differentiate an annotation in Jandex?
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: `Modifier.isInterface()` returns true even for normal interfaces, no? I don't think there is an easy way to differentiate annotations though - see https://github.com/wildfly/jandex/issues/60. I'm using this workaround: https://github.com/quarkusio/quarkus/blob/master/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/NoArgsConstructorProcessor.java#L151.

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 2]
```diff
@@ -361,6 +373,27 @@ void init() {
         return bindings;
     }
 
+    static Map<DotName, ClassInfo> findTransitiveInterceptorBindings(Collection<DotName> initialBindings,
+            IndexView index, Map<DotName, ClassInfo> result) {
+        // any annotation that contains interceptor binding annotation is effectively also an interceptor binding
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: I don't think so. It must be annotated with `@InterceptorBinding`. From the spec:
_"An interceptor binding type may declare other interceptor bindings. Interceptor bindings are transitive — an interceptor binding declared by an interceptor binding type is inherited by all components and other interceptor binding types that declare that interceptor binding type."_

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 1 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: Ah, I was wondering what was that magical trick for ;-)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java (hunk 2 above)
Comment (manovotn) [reply independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: So it seems that the spec indeed implies more strict rules whereas my implementation is lenient.
I'll revisit this on Mon and will try to alter it accordingly.

**Comment by manovotn** [comment]
@mkouba what about now?
I've move the magical constant to `io.quarkus.arc.processor.Annotations` since that seems to be helper class already and it was now used in two places.

I've altered how we look for transitive bindings - it is now aligned with CDI and only takes into consideration annotations that are bindings themselves. And I've of course added tests for that.

**Comment by manovotn** [comment]
* Reworked whole map caching, we now use `Map<DotName, List<AnnotationInstance>>`
* Interceptors themselves now allow for transitive bindings (see `TransitiveCounterInterceptor` in tests)
* We now use `AnnotationStore` while building up the map and when creating new interceptors to have up-to-date information on annotations which could have been changed via transformers (see `TransitiveInterceptionWithTransformerApplicationTest` which should verify that transformer changes are applied)

File: independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java [hunk 3]
```diff
@@ -355,12 +362,49 @@ void init() {
...
         Map<DotName, ClassInfo> bindings = new HashMap<>();
+        // Note: doesn't use AnnotationStore, this will operate on classes without applying annotation transformers
         for (AnnotationInstance binding : index.getAnnotations(DotNames.INTERCEPTOR_BINDING)) {
             bindings.put(binding.target().asClass().name(), binding.target().asClass());
         }
         return bindings;
     }
 
+    Map<DotName, Set<AnnotationInstance>> findTransitiveInterceptorBindigs(Collection<DotName> initialBindings, IndexView index,
+            Map<DotName, Set<AnnotationInstance>> result, AnnotationStore store) {
```
Comment (mkouba) [inline independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java]: There is no need to pass the `AnnotationStore` - it's a field of `BeanDeployment`.

**Review by mkouba — CHANGES_REQUESTED** [review]
I've added one minor comment but otherwise looks really good!

**Comment by manovotn** [comment]
@mkouba addressed your comment

**Review by mkouba — APPROVED** [review]

**Pull request merged**